src/features/core/metrics/errorReporting/providers/Bugsnag.js

```diff
@@ -0,0 +1,240 @@
+const GLOBAL_LIBRARY_OBJECT_NAMES = [
+  ERROR_REPORT_PROVIDER_NAME_BUGSNAG,
+  'Bugsnag',
+  BUGSNAG_LIB_INSTANCE_GLOBAL_KEY_NAME,
+  ERROR_REPORTING_SERVICE_GLOBAL_KEY_NAME,
+];
+
+// This API key token is parsed in the CI pipeline
+const API_KEY = '__RS_BUGSNAG_API_KEY__';
+
+// Errors only from Below SDKs are allowed to reach Bugsnag
+const SDK_FILE_NAMES = [
+  'rudder-analytics.min.js',
+  'rudder-analytics-staging.min.js',
+  'rudder-analytics.js',
+  ...Object.keys(configToIntNames).map((intgName) => `${configToIntNames[intgName]}.min.js`),
+  ...Object.keys(configToIntNames).map(
+    (intgName) => `${configToIntNames[intgName]}-staging.min.js`,
+  ),
+  ...Object.keys(configToIntNames).map((intgName) => `${configToIntNames[intgName]}.js`),
+];
+const BUGSNAG_ERROR_FILE_ORIGIN_KEY = 'stacktrace.0.file';
+
+const getReleaseStage = () => {
+  const host = window.location.hostname;
+  const devHosts = ['www.rs-unit-test-host.com', 'localhost', '127.0.0.1', '[::1]'];
+
+  return host && devHosts.includes(host) ? 'development' : '__RS_BUGSNAG_RELEASE_STAGE__';
+};
+
+const isValidVersion = (globalLibInstance) => {
+  // For version 7
+  let version =
+    globalLibInstance &&
+    globalLibInstance._client &&
+    globalLibInstance._client._notifier &&
+    globalLibInstance._client._notifier.version;
+
+  // For versions older than 7
+  if (!version) {
+    let tempInstance = globalLibInstance({
+      apiKey: API_KEY,
+      releaseStage: 'version-test',
+      // eslint-disable-next-line func-names, object-shorthand
+      beforeSend: function () {
+        return false;
+      },
+    });
+    version = tempInstance.notifier && tempInstance.notifier.version;
+    tempInstance = undefined;
+  }
+
+  return version && version.charAt(0) === BUGSNAG_VALID_MAJOR_VERSION;
+};
+
+const isRudderSDKError = (event) => {
+  const errorOrigin = get(event, BUGSNAG_ERROR_FILE_ORIGIN_KEY);
+
+  if (!errorOrigin || typeof errorOrigin !== 'string') {
+    return false;
+  }
+
+  const srcFileName = errorOrigin.substring(errorOrigin.lastIndexOf('/') + 1);
+  return SDK_FILE_NAMES.includes(srcFileName);
+};
+
+const enhanceErrorEventMutator = (event, metadataSource) => {
+  event.updateMetaData('source', {
+    metadataSource,
+  });
+
+  const { errorMessage } = event;
+  // eslint-disable-next-line no-param-reassign
+  event.context = errorMessage;
+
+  // Hack for easily grouping the script load errors
+  // on the dashboard
+  if (errorMessage.includes('error in script loading')) {
+    // eslint-disable-next-line no-param-reassign
+    event.context = 'Script load failures';
+  }
+
+  // eslint-disable-next-line no-param-reassign
+  event.severity = 'error';
+};
+
+const loadBugsnagSDKScript = (name) => {
+  const isNotLoaded = GLOBAL_LIBRARY_OBJECT_NAMES.every(
+    (globalKeyName) => !Object.hasOwn(window, globalKeyName),
+  );
+
+  if (isNotLoaded) {
+    ScriptLoader(name, BUGSNAG_CDN_URL, {
+      isNonNativeSDK: 'true',
+      async: 'false',
```

Comment:
Why load in sync? This additional script load might affect the app's performance indicators like FID, TTI, etc.

Reply:
Yes, better to be on the safe side. We also load all our integration SDKs async.